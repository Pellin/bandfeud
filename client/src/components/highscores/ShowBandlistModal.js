import React from 'react';
import Modal from 'react-modal';

import moment from 'moment';

import BfGoA from '../../icons/BfGoA';

Modal.setAppElement(document.getElementById('root'));

const ShowBandlistModal = props => {
  const points = [];
  if (props.showModal && props.lastRound) {
    for (let band of props.bands) {
      if (band.points) {
        points.push(band.points);
      }
    }
  }
  const goToDiscogs = band => {
    if (band.mode) return;
    if (band.id) {
      window.open(`https://www.discogs.com/artist/${band.id}`);
    } else {
      if (band.name === 'n.w.a') {
        band.name = 'n.w.a.';
      }
      const arr = band.name.split(' ');
      const string = arr.join('+');
      window.open(`https://www.discogs.com/artist/${string}`);
    }
  };
  return (
    <Modal
      isOpen={!!props.showModal}
      onRequestClose={props.closeModal}
      contentLabel="BandList"
      closeTimeoutMS={200}
      className="modal"
      overlayClassName="overlay"
    >
      <div className="modal-header">
        {props.playerInfo && (
          <>
            <div className="modal-header-left">
              {moment(props.playerInfo.date).format('DD/MM/YYYY')}{' '}
            </div>
            <div className="header-player">
              <div className="header-stopwatch">
                <BfGoA />
              </div>
              {props.playerInfo.player}
              <p>
                {props.bands.length / 2 - 0.5}{' '}
                {props.bands.length / 2 - 0.5 === 1 ? 'band' : 'bands'}
              </p>
            </div>
            <div className="modal-header-right">{props.playerInfo.score}p</div>
          </>
        )}
      </div>
      {props.lastRound && (
        <div className="modal-header">
          <div className="last-round-info">
            {points.length && points.reduce((acc, curr) => acc + curr)}p
            <p>
              {props.bands.length / 2 - 0.5}{' '}
              {props.bands.length / 2 - 0.5 === 1 ? 'band' : 'bands'}
            </p>
          </div>
        </div>
      )}
      <div className="modal-flow">
        {props.showModal &&
          props.bands.map(band => (
            <div className="band-info" key={band.mode ? band.mode : band.name}>
              <div className="band-image-container">
                {band.url ? (
                  <img
                    title={`Open Discogs page for ${band.name.toUpperCase()} `}
                    onClick={() => goToDiscogs(band)}
                    className="band-image"
                    src={band.url}
                    alt={band.name}
                  />
                ) : (
                  <div className="fail-svg">
                    <BfGoA />
                  </div>
                )}
              </div>
              <div className="band-name-container">
                {band.points ? (
                  <div className="band-points">{band.points}p</div>
                ) : band.mode ? (
                  <div className="band-points">{band.mode}</div>
                ) : (
                  <div className="no-band-points">No points</div>
                )}
                <div
                  className={
                    band.points
                      ? 'band-name'
                      : band.url
                      ? 'band-name-computer'
                      : 'band-name-failed'
                  }
                >
                  {band.name}
                </div>
              </div>
            </div>
          ))}
      </div>
    </Modal>
  );
};

export default ShowBandlistModal;
