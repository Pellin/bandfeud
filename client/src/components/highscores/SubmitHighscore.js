import React, { useState } from 'react';
import { connect } from 'react-redux';

import CrownSvg from '../../icons/CrownSvg';

import getHighscores from '../../actions/getHighscores';

export const SubmitHighscore = props => {
  const [name, setName] = useState('');
  const [highscoreSet, setHighscoreSet] = useState(false);
  const [nameSubmitted, setNameSubmitted] = useState(false);

  const onTextChange = event => {
    setName(event.target.value);
  };
  const onSubmit = async e => {
    e.preventDefault();
    setNameSubmitted(true);
    try {
      const data = {
        bands: props.bands,
        date: new Date().getTime(),
        player: name,
        score: props.score
      };
      const setHighscore = await fetch(`/api/sethighscore`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(data)
      });
      if (setHighscore.status === 200) {
        setHighscoreSet(true);
        props.onGetHighscores();
      } else {
        alert("Couldn't set highscore");
        return goToPlay();
      }
    } catch (e) {
      console.log(e.message);
    }
  };
  const goToPlay = () => {
    props.onHighscoreSet();
    props.onResetUsed();
    props.onResetScore();
    props.onResetDifficulty();
    props.onGameOver();
  };
  const goToHighscores = () => {
    props.onHighscoreSet();
    props.onResetUsed();
    props.onResetScore();
    props.onResetDifficulty();
    props.onGameOver();
    props.history.push('/highscores');
  };
  return (
    <div className="submit-view">
      {!highscoreSet ? (
        nameSubmitted ? (
          <div className="highscore-message">Setting highscore...</div>
        ) : (
          <>
            <CrownSvg className="submit-highscore-crown" />
            <div className="highscore-message">
              {props.score} points! That's a highscore.
            </div>
            <form className="submit-form" onSubmit={onSubmit}>
              <input
                className="submit-input"
                type="text"
                autoFocus
                autoCorrect="false"
                maxLength="10"
                onChange={onTextChange}
                placeholder="NAME"
                spellCheck={false}
              />
            </form>
          </>
        )
      ) : (
        <>
          <div className="highscore-message">Highscore set!</div>
          <div className="highscore-link-container">
            <div className="link-button" onClick={goToPlay}>
              âœ“
            </div>
            <CrownSvg
              className="goto-highscore-crown"
              onClick={goToHighscores}
            />
          </div>
        </>
      )}
    </div>
  );
};

const mapStateToProps = state => ({
  score: state.score,
  isHighscore: state.isHighscore
});

const mapDispatchToProps = dispatch => ({
  onGetHighscores: () => dispatch(getHighscores()),
  onHighscoreSet: () => dispatch({ type: 'HIGHSCORE:_FALSE' }),
  onResetUsed: () => dispatch({ type: 'RESET_USED' }),
  onSubmittedTrue: () => dispatch({ type: 'SUBMITTED:_TRUE' }),
  onResetScore: () => dispatch({ type: 'RESET_SCORE' }),
  onResetDifficulty: () => dispatch({ type: 'RESET_DIFFICULTY' }),
  onResetBands: () => dispatch({ type: 'RESET_BANDS' }),
  onGameOver: () => dispatch({ type: 'GAME_OVER' })
});

export default connect(mapStateToProps, mapDispatchToProps)(SubmitHighscore);
