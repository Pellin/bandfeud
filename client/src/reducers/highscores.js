export default function(state = [], action) {
  switch (action.type) {
    case 'SET_HIGHSCORES':
      return [...action.payload];
    default:
      return state;
  }
}
