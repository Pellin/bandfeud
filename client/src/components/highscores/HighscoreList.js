import React, { useState } from 'react';

import ShowBandlistModal from '../highscores/ShowBandlistModal';
import Crown from '../../icons/Crown';

const HighscoreList = ({ highscores }) => {
  const [showModal, setShowModal] = useState(false);
  const [bands, setBands] = useState([]);
  const [playerInfo, setPlayerInfo] = useState({
    index: 0,
    player: '',
    date: null,
    score: 0
  });
  const closeModal = () => {
    setShowModal(false);
    setBands([]);
  };
  const onHighscoreCLick = highscore => {
    setPlayerInfo({
      player: highscore.player,
      date: highscore.date,
      score: highscore.score
    });
    setBands(highscore.bands);
    setShowModal(true);
  };
  return (
    <>
      <div className="title-container">
        <Crown width={25} height="100%" className="crown-left" />
        <div className="highscore-title">HIGHSCORES</div>
        <Crown width={25} height="100%" className="crown-right" />
      </div>
      <table>
        <tbody className="highscore-body">
          {highscores.map(highscore => (
            <tr
              className="hs-row"
              onClick={() => onHighscoreCLick(highscore)}
              key={highscore._id}
            >
              <td className="hs-player">{highscore.player}</td>
              <td className="hs-score">{highscore.score}</td>
            </tr>
          ))}
        </tbody>
      </table>
      <ShowBandlistModal
        bands={bands}
        playerInfo={playerInfo}
        closeModal={closeModal}
        showModal={showModal}
      />
    </>
  );
};

export default HighscoreList;
