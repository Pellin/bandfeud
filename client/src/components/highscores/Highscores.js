import React, { Component } from 'react';
import { connect } from 'react-redux';

import Header from '../Header';
import HighscoreList from '../highscores/HighscoreList';

import getHighscores from '../../actions/getHighscores';

class Highscores extends Component {
  constructor(props) {
    super(props);
    this.state = {
      highscore: false
    };
  }
  componentDidMount = () => {
    if (!this.props.highscores.length) {
      this.props.onGetHighscores();
    }
  };
  render() {
    return (
      <div>
        <Header />
        <div className="highscore-container">
        
          {this.props.highscores.length ? (
            <HighscoreList highscores={this.props.highscores} />
          ) : (
            <div className="load-message">Loading highscores...</div>
          )}
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  highscores: state.highscores
});

const mapDispatchToProps = dispatch => ({
  onGetHighscores: () => dispatch(getHighscores())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Highscores);
