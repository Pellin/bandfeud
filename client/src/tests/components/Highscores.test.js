import React from 'react';
import { shallow } from 'enzyme';

import { Highscores } from '../../components/highscores/Highscores';

let highscores, onGetHighscoresSpy;

beforeEach(() => {
  highscores = [
    {
      _id: 'one',
      date: 1567247152162,
      score: 97,
      player: 'falle'
    },
    {
      _id: 'two',
      date: 1568387025126,
      score: 118,
      player: 'babs'
    },
    {
      _id: 'three',
      date: 1568447073138,
      score: 123,
      player: 'mob pete'
    },
    {
      _id: 'four',
      date: 1568734844461,
      score: 147,
      player: 'krille k'
    }
  ];
  onGetHighscoresSpy = jest.fn();
});

it('should render component correctly with highscores', () => {
  const wrapper = shallow(
    <Highscores highscores={highscores} onGetHighscores={onGetHighscoresSpy} />
  );
  
  expect(wrapper.find('HighscoreList').props().highscores.length).toBe(4);
  expect(wrapper.find('HighscoreList').props().highscores[0]).toEqual(highscores[0]);
  expect(wrapper.find('.load-message').length).toBe(0);
  expect(wrapper).toMatchSnapshot();
});

it('should render component correctly without highscores', () => {
  const wrapper = shallow(
    <Highscores highscores={[]} onGetHighscores={onGetHighscoresSpy} />
  );
  
  expect(wrapper.find('.load-message').props().children).toBe('Loading highscores...');
  expect(wrapper.find('.load-message').length).toBe(1);
  expect(wrapper).toMatchSnapshot();
});
