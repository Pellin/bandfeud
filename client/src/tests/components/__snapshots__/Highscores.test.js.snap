// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component correctly with highscores 1`] = `
<div>
  <Connect(Header) />
  <div
    className="highscore-container"
  >
    <HighscoreList
      highscores={
        Array [
          Object {
            "_id": "one",
            "date": 1567247152162,
            "player": "falle",
            "score": 97,
          },
          Object {
            "_id": "two",
            "date": 1568387025126,
            "player": "babs",
            "score": 118,
          },
          Object {
            "_id": "three",
            "date": 1568447073138,
            "player": "mob pete",
            "score": 123,
          },
          Object {
            "_id": "four",
            "date": 1568734844461,
            "player": "krille k",
            "score": 147,
          },
        ]
      }
    />
  </div>
</div>
`;

exports[`should render component correctly without highscores 1`] = `
<div>
  <Connect(Header) />
  <div
    className="highscore-container"
  >
    <div
      className="load-message"
    >
      Loading highscores...
    </div>
  </div>
</div>
`;
